### PR TITLE
Add the new account recovery settings state tree, `account-recovery/settings`

### PR DESCRIPTION
**Pull request by southp**
## Summary
This PR aims to complete the account recovery settings state tree. ~It is based on #9111 and #9173 , and thus should keep the In Progress label before those two got in.~ ( these two have been merged )

In general, it consists of the following of `account-recovery/settings` state

* The reducer and its tests
* The selectors
* Some additional polishes against the actions

## Test Plan
Run `npm run test-client client/state/account-recovery/test/` or let CircleCI do the fun.

### CODE REVIEW

**Comment by southp** [comment]
A note to myself: moving all the status flags in to individual reducers, and out from `settings`

**Comment by southp** [comment]
@omarjackman @jordwest If there are some slack times, it would be wonderful if you could review this in the following days. Thanks :)

File: client/state/account-recovery/settings/reducer.js [hunk 1]
```diff
@@ -1,33 +1,137 @@
...
+	[ ACCOUNT_RECOVERY_SETTINGS_DELETE_SUCCESS ]: ( state, { target } ) => {
+		switch ( target ) {
+			case 'phone':
+				return {
+					...state,
+					phone: null,
+				};
+			case 'email':
+				return {
+					...state,
```
Comment (jordwest) [inline client/state/account-recovery/settings/reducer.js]: `data` is a good candidate to use another level of `combineReducers` to ensure that the state tree is initialized properly, especially since you already know what all the possible properties will be. The more manual object manipulation, the harder it can be to track down strange bugs where objects don't match what you expect. `combineReducers` helps ensure a fixed state tree shape to avoid those situations.

Eg:

```js
const phone = createReducer( null, {
	[ ACCOUNT_RECOVERY_SETTINGS_FETCH_SUCCESS ]: ( state, { settings } ) => {
		return convertPhoneResponse( settings.phone )
	},
	[ ACCOUNT_RECOVERY_SETTINGS_UPDATE_SUCCESS ]: ( state, { target, value } ) => {
		if ( target === 'phone' ) {
			return convertPhoneResponse( settings.phone );
		}
		return state;
	},
	[ ACCOUNT_RECOVERY_SETTINGS_DELETE_SUCCESS ]: ( state, { target, value } ) => {
		if ( target === 'phone' ) {
			return null;
		}
		return state;
	},
} );

// ...

export default combineReducers( {
	data: combineReducers( { phone, email } ),
	isUpdating,
	isDeleting,
} );
```

It would even be possible to take it a step further and build a function to generate the reducer:

```js
const createDataReducer = ( field, converter ) => {
	return createReducer( null, {
		[ ACCOUNT_RECOVERY_SETTINGS_FETCH_SUCCESS ]: ( state, { settings } ) => {
			return converter( settings[ field ] )
		},
		[ ACCOUNT_RECOVERY_SETTINGS_UPDATE_SUCCESS ]: ( state, { target, value } ) => {
			if ( target === field ) {
				return converter( settings[ field ] );
			}
			return state;
		},
		[ ACCOUNT_RECOVERY_SETTINGS_DELETE_SUCCESS ]: ( state, { target, value } ) => {
			if ( target === field ) {
				return null;
			}
			return state;
		},
	} );
}

// ...

export default combineReducers( {
	data: combineReducers( {
		phone: createDataReducer( 'phone', convertPhoneResponse ),
		email: createDataReducer( 'email', convertEmailResponse ),
	} ),
	isUpdating,
	isDeleting,
} );
```

File: client/state/account-recovery/test/reducer.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	} );
+
+	it( 'should unset isFetchingSettings flag on success.', () => {
+		const state = reducer( undefined, {
+			type: ACCOUNT_RECOVERY_SETTINGS_FETCH_SUCCESS,
+			settings: {
+				email: '',
+				email_validated: true,
+				phone: {},
+				phone_validated: true,
```
Comment (southp) [inline client/state/account-recovery/test/reducer.js]: The meaning behind these two flags is that we will require validating of the ownership and accessibility of recovery emails and phones. In case of recovery phones, once a user adds / updates ones recovery phone, it is considered "unvalidated" and `phone_validated` will be `false`. After going through the validation process, it is "validated" and `phone_validated` will be `true`.

If I understand your meaning correctly, I think it is more of the former, is it?

Comment (jordwest) [inline client/state/account-recovery/test/reducer.js]: I think so, I guess I'm confused about the meaning of `validated`.

For example, if it fails validation, I'd expect `validated === true` but `valid === false`

File: client/state/account-recovery/settings/reducer.js [hunk 3]
```diff
@@ -1,33 +1,137 @@
...
+
+	return {
+		countryCode: country_code,
+		countryNumericCode: country_numeric_code,
+		number: number,
+		numberFull: number_full,
+	};
+};
+
+const convertEmailResponse = ( emailResponse ) => {
```
Comment (omarjackman) [inline client/state/account-recovery/settings/reducer.js]: We could shorten this to

```js
const convertEmailResponse = emailResponse => emailResponse || '';
```

Comment (omarjackman) [inline client/state/account-recovery/settings/reducer.js]: Just noticing how small this function is makes me wonder if we really need it. Are there plans to add more logic here?

File: client/state/account-recovery/settings/reducer.js (hunk 1 above)
Comment (omarjackman) [inline client/state/account-recovery/settings/reducer.js]: I think `createDataReducer` makes this look easy and allows you to cut down on the repetition but something feels wrong to me about creating reducers in this way. For the sake of keeping it simple I'd vote for the first approach you mentioned.

File: client/state/account-recovery/settings/selectors.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+export const isAccountRecoveryPhoneValidated = ( state ) => {
+	return isAccountRecoverySettingsReady( state ) ? getSettingsData( state ).phoneValidated : false;
+};
+
+export const isAccountRecoveryEmailValidated = ( state ) => {
+	return isAccountRecoverySettingsReady( state ) ? getSettingsData( state ).emailValidated : false;
+};
+
+export const getAccountRecoveryEmail = ( state ) => {
+	return isAccountRecoverySettingsReady( state ) ? getSettingsData( state ).email : '';
```
Comment (omarjackman) [inline client/state/account-recovery/settings/selectors.js]: `isAccountRecoverySettingsReady` is checking if `state.accountRecovery.settings.data` is null so that might actually not be a problem

File: client/state/account-recovery/settings/selectors.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+export const isAccountRecoverySettingsReady = ( state ) => {
+	return null != state.accountRecovery.settings.data;
```
Comment (omarjackman) [inline client/state/account-recovery/settings/selectors.js]: Maybe we should use `getSettingsData` here

File: client/state/account-recovery/settings/reducer.js [hunk 6]
```diff
@@ -1,33 +1,137 @@
...
+				return { ...state };
+		}
+	},
+
+	[ ACCOUNT_RECOVERY_SETTINGS_DELETE_SUCCESS ]: ( state, { target } ) => {
+		switch ( target ) {
+			case 'phone':
+				return {
+					...state,
+					phone: null,
```
Comment (omarjackman) [inline client/state/account-recovery/settings/reducer.js]: Should we also be clearing out `phoneValidated`?

**Comment by omarjackman** [comment]
@southp I left a few comments but I agree with @jordwest [comments](https://github.com/Automattic/wp-calypso/pull/9208/files#r93232162) concerning your reducers

File: client/state/account-recovery/test/selectors.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Internal dependencies
+ */
+import {
+	isFetchingAccountRecoverySettings,
+} from '../selectors';
+
+describe( '#account-recovery selector isFetchingAccountRecoverySettings:', () => {
+	it( 'should return the correct property.', () => {
```
Comment (southp) [inline client/state/account-recovery/test/selectors.js]: Fix in 8e6546a92fe5b1203843322d915922e47fa9dfc5.

File: client/state/account-recovery/settings/reducer.js (hunk 3 above)
Comment (southp) [inline client/state/account-recovery/settings/reducer.js]: Nice and concise :)
There is no plan to add more logic, but as refactoring the `data` reducer in 75b2fe39319f01ee8256e780b8c2d1ad5868f46f, it now uses in two places so I think it still worths a function.

File: client/state/account-recovery/settings/reducer.js (hunk 1 above)
Comment (southp) [inline client/state/account-recovery/settings/reducer.js]: Love the suggestion! Like @omarjackman mentioned, I choose to keep it simple in 75b2fe39319f01ee8256e780b8c2d1ad5868f46f.

File: client/state/account-recovery/test/reducer.js (hunk 2 above)
Comment (southp) [inline client/state/account-recovery/test/reducer.js]: I see your point now. It might be not the best name of it, but I have no better idea yet and it would require to patch the endpoint. Is it okay to you to keep it this way?

**Comment by southp** [comment]
@jordwest @omarjackman Thanks for the review. The comments above are all fixed or answered. It would be wonderful if you can revisit it. Thanks :)

File: client/state/account-recovery/settings/selectors.js (hunk 5 above)
Comment (southp) [inline client/state/account-recovery/settings/selectors.js]: Nice catch. Fixed by 75b2fe39319f01ee8256e780b8c2d1ad5868f46f

File: client/state/account-recovery/settings/reducer.js (hunk 6 above)
Comment (southp) [inline client/state/account-recovery/settings/reducer.js]: True. It is cleaned along with the changes in 38e90e6f9510df606a803c09d8ee5acaa4d34068

File: client/state/account-recovery/test/reducer.js (hunk 2 above)
Comment (jordwest) [inline client/state/account-recovery/test/reducer.js]: It's okay as is, was just a suggestion :)

File: client/state/account-recovery/selectors.js [hunk 8]
```diff
@@ -0,0 +1 @@
+export const isFetchingAccountRecoverySettings = ( state ) => state.accountRecovery.isFetchingSettings;
```
Comment (southp) [inline client/state/account-recovery/selectors.js]: Thanks for pointing this out. I will make this migration as a separate janitorial PR in the near future :)
